pkg/reconciler/apiresource/controller.go

```diff
@@ -0,0 +1,373 @@
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	if obj == nil {
+		return
+	}
+
+	theType, gvr, oldMeta, newMeta, oldStatus, newStatus := toQueueElementType(oldObj, obj)
+	var theAction QueueElementAction
+	var deletedObject interface{}
+
+	switch action {
+	case "Add":
+		theAction = Created
+	case "Update":
+		if oldMeta == nil {
+			theAction = Created
+			break
+		}
+
+		if oldMeta.GetResourceVersion() == newMeta.GetResourceVersion() {
+			return
+		}
+
+		if oldMeta.GetGeneration() != newMeta.GetGeneration() {
+			theAction = SpecChangedAction
+			break
+		}
+
+		if !equality.Semantic.DeepEqual(oldStatus, newStatus) {
+			theAction = StatusOnlyChangedAction
+			break
+		}
+
+		if !equality.Semantic.DeepEqual(oldMeta.GetAnnotations(), newMeta.GetAnnotations()) ||
+			equality.Semantic.DeepEqual(oldMeta.GetLabels(), newMeta.GetLabels()) {
+			theAction = AnnotationOrLabelsOnlyChanged
+			break
+		}
+		// Nothing significant changed. Ignore the event.
+		return
+	case "Delete":
+		theAction = Deleted
+		deletedObject = obj
+	}
+
+	c.queue.AddRateLimited(queueElement{
+		theAction:     theAction,
+		theType:       theType,
+		theKey:        key,
+		gvr:           gvr,
+		clusterName:   newMeta.GetClusterName(),
+		deletedObject: deletedObject,
+	})
+}
+
+func (c *Controller) Start(numThreads int) {
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(c.startWorker, time.Second, c.stopCh)
+	}
+	klog.Info("Starting workers")
+}
+
+// Stop stops the controller.
+func (c *Controller) Stop() {
+	klog.Info("Stopping workers")
+	c.queue.ShutDown()
+	close(c.stopCh)
+}
+
+// Done returns a channel that's closed when the controller is stopped.
+func (c *Controller) Done() <-chan struct{} { return c.stopCh }
+
+func (c *Controller) startWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(queueElement)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	err := c.process(key)
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key queueElement) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		klog.Info("Successfully reconciled", key)
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if errors.IsRetryable(err) || num < 5 {
+		klog.Info("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Infof("Dropping key %q after failed retries: %v", key, err)
+}
+
+func RegisterClusterCRD(cfg *rest.Config) error {
```

Comment:
I believe this method can be deleted, right?

Reply:
Fixed in commit 6a4dcd07a4a268ef7dd0158950e2a1c8452b9eb3